Add pta self test for firmware handoff

raymo200915 opened the pull request:
Fix a pointer casting bug in transfer_list_add().

Add self tests for transfer list.
Adapt CFG_TRANSFER_LIST with its dependencies and add CFG_TRANSFER_LIST_TEST.

<!--
    If you are new to submitting pull requests to OP-TEE, then please have a
    look at the list below and tick them off before submitting the pull request.

    1. Read our contribution guidelines:
         https://optee.readthedocs.io/en/latest/general/contribute.html

    2. Read the contribution section in Notice.md and pay extra attention to the
       "Developer Certificate of Origin" in the contribution guidelines.

    3. You should run checkpatch preferably before submitting the pull request.

    4. When everything has been reviewed, you will need to squash, rebase and
       add tags like `Reviewed-by`, `Acked-by`, `Tested-by` etc. More details
       about this can also be found on the link provided above.

    NOTE: This comment will not be shown in the pull request, so no harm keeping
    it, but feel free to remove it if you like.
-->

jenswi-linaro commented on core/kernel/transfer_list.c:
```diff
@@ -497,7 +497,7 @@ struct transfer_list_entry *transfer_list_add(struct transfer_list_header *tl,
 
 	if (data)
-		memmove(tl_e + tl_e->hdr_size, data, data_size);
+		memmove((void *)tl_e + tl_e->hdr_size, data, data_size);
```
Please replace `void` with `uint8_t` since we don't allow pointer arithmetic with void pointers.

raymo200915 replied:
Fixed by using `transfer_list_entry_data()` instead.

jenswi-linaro commented on mk/config.mk:
```diff
@@ -513,11 +513,10 @@ endif
+$(eval $(call cfg-enable-all-depends,CFG_TRANSFER_LIST, \
+	 CFG_DT CFG_EXTERNAL_DT CFG_MAP_EXT_DT_SECURE))
+
+CFG_TRANSFER_LIST_TEST ?= $(call cfg-all-enabled,CFG_TRANSFER_LIST)
```
`CFG_TRANSFER_LIST_TEST ?= $(CFG_TRANSFER_LIST)` is enough.

raymo200915 replied:
Fixed.

jenswi-linaro commented on core/pta/tests/sub.mk:
```diff
@@ -6,3 +6,4 @@ cflags-misc.c-y += -fno-builtin
 srcs-y += mutex.c
 srcs-y += aes_perf.c
 srcs-$(CFG_DT_DRIVER_EMBEDDED_TEST) += dt_driver_test.c
+srcs-$(CFG_TRANSFER_LIST_TEST) += tl.c
```
`tl.c` is a bit short. Please make it `transfer_list.c`.

raymo200915 replied:
Fixed.

raymo200915 commented on core/pta/tests/transfer_list.c:
```diff
@@ -0,0 +1,247 @@
+	struct transfer_list_entry *te1, *te2, *te3;
+	TEE_Result ret;
+
+	va_tl = memalign(TL_ALIGNMENT_FROM_ORDER(TRANSFER_LIST_INIT_MAX_ALIGN),
```
But `nex_phys_mem_core_alloc()` has not an arg for a user specified alignment, is there any other alternatives I can take?

jenswi-linaro replied:
You get full pages, so the alignment shouldn't be an issue. Please use `phys_mem_core_alloc ()` instead of `nex_phys_mem_core_alloc ()`. I made a typo above.

raymo200915 replied:
Done.

jforissier commented on core/pta/tests/invoke.c:
```diff
@@ -433,6 +433,10 @@ static TEE_Result invoke_command(void *pSessionContext __unused,
+#if defined(CFG_TRANSFER_LIST_TEST)
+	case PTA_INVOKE_TESTS_CMD_TRANSFER_LIST_TESTS:
+		return core_transfer_list_tests(nParamTypes, pParams);
+#endif
```
There is a problem when `CFG_TRANSFER_LIST_TEST` is disabled: `xtest` (as it is currently in https://github.com/OP-TEE/optee_test/pull/788) will report a failure. One option is to return success (see `core_dt_driver_tests()`). But
TBH I'd rather we return `TEE_ERROR_NOT_SUPPORTED` and then have `xtest` print the proper warning, I think it makes more sense.

raymo200915 replied:
In case `CFG_TRANSFER_LIST_TEST` is disabled, how about to print warnings in `core_dt_driver_tests()` but still return success? This can avoid having a new `Do_ADBG_#` function in optee_test.

raymo200915 replied:
Please see commit https://github.com/OP-TEE/optee_test/pull/788/commits/388bd391ad7d63efd0a1ed8cc5c15e7b0f28eec0
I have added macros and helper function to avoid returning error for not supported tests.

jforissier replied:
Good, thank you.

raymo200915 commented on core/pta/tests/transfer_list.c:
```diff
@@ -0,0 +1,247 @@
+	vaddr_t old_tl_ev = (vaddr_t)tl + tl->size;
+	struct transfer_list_entry *tl_e;
+
+	if (align > tl->alignment)
```
Good idea. Done.

raymo200915 commented on core/pta/tests/transfer_list.c:
```diff
@@ -0,0 +1,247 @@
+			      const void *data, uint8_t align,
+			      struct transfer_list_entry **tle)
+{
+	uint8_t *te_dat;
```
Done.

jenswi-linaro commented on core/pta/tests/transfer_list.c:
```diff
@@ -0,0 +1,242 @@
+static TEE_Result transfer_list_tests(void)
+{
+	struct transfer_list_header *tl = NULL;
+	struct transfer_list_entry *te1 = NULL, *te2 = NULL, *te3 = NULL;
```
Please declare one local variable per line, like:
``` C
struct transfer_list_entry *te1 = NULL;
struct transfer_list_entry *te2 = NULL;
...
```

raymo200915 replied:
Done.

jforissier commented:
For "core: kernel: fix bug in transfer_list_add" with the fixup:
Please add parentheses in the commit subject: `transfer_list_add()`, then:
`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

jforissier commented:
For commit "core: pta: add self tests for transfer list" with the fixups and Jens' comment addressed:
`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

etienne-lms commented on mk/config.mk:
```diff
@@ -513,11 +513,10 @@ endif
+$(eval $(call cfg-enable-all-depends,CFG_TRANSFER_LIST, \
+	 CFG_DT CFG_EXTERNAL_DT CFG_MAP_EXT_DT_SECURE))
+
+CFG_TRANSFER_LIST_TEST ?= $(CFG_TRANSFER_LIST)
```
A short description would be nice. Maybe mention that `CFG_TRANSFER_LIST_TEST` is not effective when `CFG_ENABLE_EMBEDDED_TESTS` is disabled.

etienne-lms commented on mk/config.mk:
```diff
@@ -513,11 +513,10 @@ endif
+$(eval $(call cfg-enable-all-depends,CFG_TRANSFER_LIST, \
+	 CFG_DT CFG_EXTERNAL_DT CFG_MAP_EXT_DT_SECURE))
+
+CFG_TRANSFER_LIST_TEST ?= $(CFG_TRANSFER_LIST)
```
A short description would be nice. Maybe tell that `CFG_ENABLE_EMBEDDED_TESTS` needs to be enabled to have the test embedded.

etienne-lms replied:
> I guess you mean CFG_TEE_CORE_EMBED_INTERNAL_TESTS?

Yes, sorry.

> I can replace it with:
> (...)

Sound good to me, still with a short description for `CFG_TRANSFER_LIST_TEST`, like:
`# Embed transfer list support self test when enabled`

raymo200915 replied:
Done.

raymo200915 commented on core/pta/tests/transfer_list.c:
```diff
@@ -0,0 +1,242 @@
+	    tl->alignment != TRANSFER_LIST_INIT_MAX_ALIGN ||
+	    tl->size != sizeof(*tl) ||
+	    tl->max_size != TEST_TL_MAX_SIZE ||
+	    tl->flags != TL_FLAGS_HAS_CHECKSUM) {
```
Added.

raymo200915 commented:
All comments are addressed.
All commits are squashed and rebased.
Added tag `Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>` to commit "core: kernel: fix bug in transfer_list_add()" and `Acked-by: Jerome Forissier <jerome.forissier@linaro.org>` to commit "core: pta: add self tests for transfer list"

jenswi-linaro reviewed (commented):
`Acked-by: Jens Wiklander <jens.wiklander@linaro.org>`

etienne-lms commented on core/pta/tests/transfer_list.c:
```diff
@@ -0,0 +1,245 @@
+		tl_e = transfer_list_add(tl, tag_id, data_size, data);
+	else
+		tl_e = transfer_list_add_with_align(tl, tag_id, data_size, data,
+						    align);
```
Nitpicking: calling only `transfer_list_add_with_align(..., align)` should be enough.

raymo200915 replied:
Here is just to test different API entries.

etienne-lms commented on core/pta/tests/transfer_list.c:
```diff
@@ -0,0 +1,245 @@
+	new_te_next = transfer_list_next(tl, tl_e);
+
+	/* skip the inserted void entry if it exists */
+	if (new_te_next->tag_id == TL_TAG_EMPTY)
```
Suggestion in case there are several empty slot + in case there is no next entry:
```patch
-	if (new_te_next->tag_id == TL_TAG_EMPTY)
+	while (new_te_next && new_te_next->tag_id == TL_TAG_EMPTY)
```

Bythe way, how come `new_te_next` is not NULL when this function is called by `transfer_list_tests()` at line 215 to change `TEST_TE3_ID` data size.

raymo200915 replied:
Good question. When expending the data region of an entry, it tries to ending up with the address that fits the potential 'next' entry with the data alignment required, even when it is the last entry - That is the case you see in line 215, a void entry was added at the tail.
It is not bad but I think it is fair enough to check if it is the last one and avoid to filling the gap with the void entry.
I will add a small fix.

etienne-lms reviewed (commented):
`Reviewed-by: Etienne Carriere <etienne.carriere@foss.st.com>` for commit
"core: kernel: fix bug in transfer_list_add()".

Nitpicking comment/question for commit "core: pta: add self tests for transfer list". Otherwise LGTM.

jforissier reviewed (commented):
For commit "core: kernel: remove the last appended void transfer entry":

`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

raymo200915 commented:
All tags are added.
